[LTD-755] User Feedback

alixedi opened the pull request:
This does 3 things -

1. Adds user email to the session. This is used in the feedback loop.
2. Adds a `feedback` app with 2 views. The first view renders a feedback form. The second view says thank you. In between the 2 views, user feedback (along with user email) is sent to a specified email inbox for triage via Notify.
3. Adds a Beta banner with a link to the feedback.

The tests cover most code paths. The only assumption is for `send_email_notification` to throw a subclass of `RequestException` when it fails. I have checked this. It does. 

I haven't included any tests for the beta banner. This is because I have chosen to put it in the template without any conf switches. I feel that it is unlikely that we will need to remove it quickly i.e. with just a config change instead of a PR.

P.S. I have gone off-spec in one place. I haven't put the "<x> characters remaining" thing under the feedback text area. I don't think it is a blocker for pvt. beta but let me know if you think I am wrong.

![image](https://user-images.githubusercontent.com/3349430/115567240-f304f980-a2b2-11eb-99d8-d28f8f13ff2f.png)

r4vi commented on unit_tests/caseworker/feedback/test_views.py:
```diff
@@ -0,0 +1,51 @@
+    assert response.status_code == 200
+
+
+@mock.patch("core.feedback.notify.client")
```
for consistency with the codebase maybe it should be a pytest fixture rather than using @mock.patch directly? example: https://github.com/uktrade/lite-frontend/blob/ca41df83ae086c0f6dff23c6c7409eeecdac56a1/unit_tests/conftest.py#L18

alixedi replied:
Err, for some reason, I cannot get the `mock_notify_client.send_email_notification.assert_called_once_with` to pass when I use a `pytest` fixture. I can try and figure out unless you are easy with this.

r4vi replied:
I'm easy, I actually prefer it this way!

r4vi commented on core/feedback/urls.py:
```diff
@@ -0,0 +1,8 @@
+
+urlpatterns = [
+    path("", FeedbackView.as_view(), name='feedback'),
+    path("/thanks", get_thanks, name='thanks'),
```
no trailing slash? feels weird seeing this in a django app

saruniitr commented on Pipfile:
```diff
@@ -57,6 +57,7 @@ crispy-forms-gds = "~=0.2"
 django-health-check = "~=3.14.3"
 django-allow-cidr = "~=0.3"
 django-redis = "~=4.12"
+notifications-python-client = "*"
```
I think we also commit `Pipfile.lock` which is missing in the PR

r4vi commented on caseworker/templates/layouts/base.html:
```diff
@@ -85,6 +85,20 @@
+				beta
+				</strong>
+				<span class="govuk-phase-banner__text">
+				This is a new service – your <a class="govuk-link" href="{% url 'feedback' %}" target="_blank">feedback</a> will help us to improve it.
```
nice using `target="_blank"`

eadpearce commented on core/feedback/forms.py:
```diff
@@ -0,0 +1,5 @@
+
+
+class FeedbackForm(forms.Form):
+    feedback = forms.CharField(label="", widget=forms.Textarea(attrs={"maxlength": "500",}),)
```
Ideally we should have a label for all form fields for accessibility. I believe the gov uk design system has a `visually-hidden` css class or similar we could use to hide the text for sighted users

eadpearce commented on exporter/templates/layouts/base.html:
```diff
@@ -84,6 +84,17 @@
 		{% endblock %}
 
 		<div class="govuk-width-container">
+			<!-- Beta Banner -->
```
Since this is also used in caseworker/templates/layouts/base.html we could split it out into a shared template though you'd probably have to tweak some stuff in settings so django knows where to look for it

alixedi replied:
That is a good point. I thought about it but decided to not do it because the beta banner is meant to be removed at some point. Doing it this way means it is significantly easier to remove - materially and from a cognitive load perspective.

alixedi commented:
I added tests for the exporter app, ran into an issue and fixed it.

I have also added the extra env vars to `example.<x>.env` files.

Finally, I have put a label on the feedback form.
